design/sequence/scheduler-low.puml

```diff
@@ -0,0 +1,86 @@
+@startuml
+
+box "Central Server" #Lightblue
+    participant ":Database"
+    participant ":Scheduler"
+    participant "job:Job"
+    participant ":Dispatcher"
+end box
+
+
+":Database" -> ":Scheduler" ++: reschedule()
+":Scheduler" -> ":Database" ++: get_changed_job()
+":Database" --> ":Scheduler" --: job: Job
+activate "job:Job"
+
+
+break job.get_status()==JobStatus.KILLED||\n job.get_status()==JobStatus.FINISHED
+    ":Scheduler" -> "job:Job" ++:check_was_running()
+    "job:Job" --> ":Scheduler" --: was_running: bool
+
+    alt was_running
+        alt job.get_status()==JobStatus.KILLED
+            ":Scheduler" -> ":Dispatcher" ++: kill_job(job)
+            ":Dispatcher" --> ":Scheduler" --
+        end
+        ":Scheduler" -> ":Database" ++: get_next_job(job.wc_id)
+        ":Database" --> ":Scheduler" --: new_job: Job
+        ":Scheduler" -> ":Dispatcher" ++ : dispatch_job(new_job)
+        ":Dispatcher" --> ":Scheduler" --:Message
+        ":Scheduler" -> ":Database"++: set_job_status(new_job, JobStatus.RUNNING)
+        ":Database" --> ":Scheduler" --
+    end
+end
+
+":Scheduler" -> ":Scheduler" ++: find_wcs(job)
+":Scheduler" -> ":Database" ++: get_wc_list()
+":Database" --> ":Scheduler" --: wc_list:list[str]
+":Scheduler" --> ":Scheduler" --: suitable_wcs:list[str]
+
+":Scheduler" -> "job:Job" ++: set_wc_ids(suitable_wcs)
+"job:Job" --> ":Scheduler" --:
+
+
+":Scheduler" -> ":Database" ++: check_idle(suitable_wcs)
+":Database" --> ":Scheduler" --: wc_id:str
+
+":Scheduler" -> "job:Job" ++: calc_cost()
+"job:Job" --> ":Scheduler" --: cost:int
+":Scheduler" -> ":Database" ++ : set_cost(job, cost)
+":Database" --> ":Scheduler" -- 
```

Comment:
Yes, I meant the effective priority, we'll fix that.

>Effective priority needs to be calculated before the Scheduler loops over all Jobs.
That's what we did, the effective priority is calculated right after a job is added. Then it's calculated again when we call the 'resort' method (since the effective priority changes over time).

Reply:
> That's what we did, the effective priority is calculated right after a job is added. Then it's calculated again when we call the 'resort' method (since the effective priority changes over time).

There is no need to specially calculate priority after job is added, you can just do it in the resort().

Reply:
I thought it would be useful to add a "effective_priority" attribute to a job. resort() sorts the list by that attribute, it's pretty much the same, I'll change if needed.

Reply:
>That's what we did, the effective priority is calculated right after a job is added. Then it's calculated again when we call the 'resort' method (since the effective priority changes over time).

The effective priority needs be calculated **in the very beginning before any Jobs are being scheduled or dispatched**.
After that the Scheduler needs to loop over the available Jobs and try to schedule them.
For this process the Scheduler **must** have the effective priority of all Jobs.
In your sequence diagram the effective priority is being calculated **after** you have already scheduled and dispatched Jobs.